hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java

```diff
@@ -287,6 +301,11 @@ public TransactionContext preAppendTransaction(TransactionContext trx)
     final OMRequest request = (OMRequest) trx.getStateMachineContext();
     OzoneManagerProtocolProtos.Type cmdType = request.getCmdType();
 
+    if (readonlyMode.get()) {
```

Comment:
Why are we checking this in so many places? Prepare does a similar thing with just a pre-append check, since all write requests must go through there. I actually don't recall the difference between preAppend and startTransaction to know which is better for the use case in this change.

Reply:
1. startTransaction() is called out of synchronized lock by ratis. On exception, it return exception to client.
2. preAppend Transaction() is called after having synchronized lock, and with success, create log entry and sync to all node in ratis ring. On Exception, it also causes dropping itself as leader as logic.

- startTransaction() check immediately return failure.
- preAppend() check is to avoid continue of already triggered request waiting on lock. This is for those request waiting for previous request to complete.

So added both places for user request.

Reply:
I don't quite follow. Is this a performance improvement? I don't think we need to worry about that when every request in this state will fail. It would be simpler just to check in preAppend.